backend/bitboxbase/updater/updater.go

```diff
@@ -83,7 +108,150 @@ func (updater *Updater) Connect(address string, bitboxBaseID string) error {
+	cipherSuite := noise.NewCipherSuite(noise.DH25519, noise.CipherChaChaPoly, noise.HashSHA256)
+	keypair := updater.configGetAppNoiseStaticKeypair()
+	if keypair == nil {
+		updater.log.Info("noise static keypair created")
+		kp, err := cipherSuite.GenerateKeypair(rand.Reader)
+		if err != nil {
+			return errors.New("unable to generate a new noise keypair for the wallet app communication with the BitBox Base")
+		}
+		keypair = &kp
+		if err := updater.configSetAppNoiseStaticKeypair(keypair); err != nil {
+			updater.log.WithError(err).Error("could not store app noise static keypair")
+			// Not a critical error, ignore.
+		}
+	}
+	handshake, err := noise.NewHandshakeState(noise.Config{
+		CipherSuite:   cipherSuite,
+		Random:        rand.Reader,
+		Pattern:       noise.HandshakeXX,
+		StaticKeypair: *keypair,
+		Prologue:      []byte("Noise_XX_25519_ChaChaPoly_SHA256"),
+		Initiator:     true,
+	})
+	if err != nil {
+		return errors.New("failed to generate a new noise handshake state for the wallet app communication with the BitBox Base")
+	}
+	//Check the websocket connection
+	err = client.WriteMessage(1, []byte(opICanHasHandShaek))
+	if err != nil {
+		return errors.New("unable to write BitBox Base Handshake request to websocket")
+	}
+	_, responseBytes, err := client.ReadMessage()
+	responseSuccess := []byte("ACK")
+	if err != nil {
+		return errors.New("unable to read BitBox Base Handshake response from websocket")
+	}
+	if string(responseBytes) != string(responseSuccess) {
+		return errors.New("no ACK received from BitBox Base as response to hanshake request")
+	}
+
+	// Do 3 part noise 'XX' handshake.
+	msg, _, _, err := handshake.WriteMessage(nil, nil)
+	if err != nil {
+		return errors.New("noise failed to write the first handshake message")
+	}
+	err = client.WriteMessage(websocket.TextMessage, msg)
+	if err != nil {
+		return errors.New("the websocket failed writing the first noise handshake message")
+	}
+	_, responseBytes, err = client.ReadMessage()
+	if err != nil {
+		return errors.New("the websocket failed reading the second noise handshake message")
+	}
+	_, _, _, err = handshake.ReadMessage(nil, responseBytes)
+	if err != nil {
+		return errors.New("noise failed to read the second handshake message")
+	}
+	msg, updater.receiveCipher, updater.sendCipher, err = handshake.WriteMessage(nil, nil)
+	if err != nil {
+		return errors.New("noise failed to write the third handshake message")
+	}
+	err = client.WriteMessage(websocket.TextMessage, msg)
+	if err != nil {
+		return errors.New("the websocket failed writing the third handshake message")
+	}
+
+	// Check if the user already authenticated the channel binding hash
+	updater.baseNoiseStaticPubkey = handshake.PeerStatic()
+	pairingVerificationRequiredByApp := !updater.configContainsBitBoxBaseStaticPubkey(
+		updater.baseNoiseStaticPubkey)
+	if pairingVerificationRequiredByApp {
+		msg = []byte(opICanHasPairinVerificashun)
+	}
```

Comment:
if clause should cover all the next lines about verifying the hash. otherwise you are re-sending the previous msg, which is garbage by now.

Reply:
This is the intention, either the base will receive garbage, or 'v'. I'd also like to do it strictly sequentially, where both sides have to wait for each other to answer, so I will change the order here.

Reply:
sending garbage doesn't sound so good :) harder to reason about and potentially fragile, e.g. the garbage could start with 'v', etc.

Please do attempt to use the same handshake as bitbox02, a lot of thought has been put into it to make all edge cases work, and having one protocol means less effort, less audit, less reasoning about all cases, and so on.